feat(other-profile-backend): implement other profile backend and follow feature

abaur3 opened the pull request:
# What Changes

This pull request completes the implementation (backend-side too) of the other user profile feature, including the follow feature. Now users are able to navigate to other user profiles, data are really fetched from the remote Firebase repository, and users are able to follow others. By clicking on the `Follow` button, the other user `subscribers` count increments (not only locally, but now also remotely) and the currently signed-in user (let's call it current user for simplicity) `subscriptions` count increments. Moreover, the other user's `uid` is added to the current user's `followingList`. If the Follow action fails, an error is shown under the Follow button. 

## Key Implementations :

### Repository and Data model refactoring (support for other profiles)
- Refactored profile repository interface: now methods used to fetch data related to a profile accept a user `uid`, in order to be able to get and observe not only the current user profile, but also the other user profile; in particular:
    - Renamed methods `observeProfile()` and `getProfile()` to `observeCurrentProfile()` and `getCurrentProfile()` respectively, so that they’re specific to the current user
    - Added methods that specifically relate to other users: `observeProfile(uid: String)` and `getProfile(uid: String)`
- Implemented all four methods in `ProfileRepositoryFirebase`, as well as a `requireCurrentUid()` helper used to return the UID of the current authenticated user, or throw if none exists. `observeCurrentProfile()` / `getCurrentProfile()` simply call `observeProfile()` and `getProfile()` respectively, using the `uid` returned by `requireCurrentUid()`
- Updated UI state models:
    - In `SelfProfileUiState`: renamed `followers`/`following` to `subscribers`/`subscriptions` for clarity; added a `followingList` with the `uid`s of followed users
    - In `OtherProfileUiState`: renamed `isFollowing` to `isCurrentUserFollowing` for clarity.
- Updated `OtherProfileViewModel` to observe real repository data: removed the previous mocked logic and snapshot field, and changed `init` to observe real profile Flows from the repository.

### Follow/Unfollow Backend implementation
- Added two new repository actions: `followUser(uid: String`) and `unfollowUser(uid: String)`, and implemented these in `ProfileRepositoryFirebase` and `FakeProfileRepository` (stub implementations for the latter). 
- Updated `OtherProfileViewModel.onFollow()` so that it calls the repository’s real follow/unfollow methods; removed the previous manual incrementing/decrementing of `subscribers` count: now UI is fully driven by backend updates
- Added `errorMessage` to `OtherProfileUiState` and wired it so that errors from follow/unfollow operations can surface in the UI (displayed under the follow button in red), and are cleared when the operation succeed. 
- Implemented a Cloud Function `followUser`: the attempt to update both the current user’s and the other user’s profile documents (in order to update the current user’s `followingList`, `subscriptions` count and the other user’s `subscribers` count) directly in the app failed because Firestore security rules forbid clients from modifying another user’s document; this is why I decided to delegate this logic to a Cloud Function, which runs with Admin privileges. I finally exported the function in `functions/src/index.ts` and deployed it
- The Cloud Function `followUser` is called by the method `callFollowFunction` in `ProfileRepositoryFirebase`

### Updating tests according to the new architecture
- `OtherProfileViewModelTest`
    - Changed some tests so they work with the new Flow based repository.
    - Added mocks for `FirebaseAuth`, `ImageStorageRepository`, and `StorageService`
    - Added a small `FollowToggleTestRepository` used only for tests
    - Updated the assertions so they wait for the Flow to emit new values instead of checking state immediately.
    - Used `advanceUntilIdle` to make coroutine timing predictable.
- `ProfileScreenTest`:
    - Updated the UI tests to follow the new backend based logic.
    - Replaced the old test for follower count with a new one that checks the count only after the backend emits an update.
    - Instead of modifying UI state directly, tests now track follow requests using a simple counter.
    - Passed the `errorMessage` field through `OtherProfileConfig` in tests that need it.
- `FakeProfileRepository`: added basic stub versions of `followUser` and `unfollowUser` so that tests compile
- New test class: `ProfileRepositoryFirebaseFollowTest`: it covers the new follow/unfollow feature:
    - verifies correct payload sent to the Cloud Function
    - checks rejection of blank or self UIDs
    - asserts that callable exceptions are propagated
    - ensures the follow/unfollow flags are correctly set

- `ProfileScreenTest`:
    - Added tests for follow errors, fake profile repo, and route behavior
    - Tested that the follow error shows only when there’s a message.
    - Added a fake other profile repository so tests can control follow/unfollow behavior.
    - Added route tests to check remote profile display, follow state changes, and the go-back callback.


### ci and build configuration changes
- Added `firebase-functions-ktx` to the app’s Gradle build (and placed Gradle dependencies in the `libs.versions.toml` file)
- Added a CI step to install and build Cloud Functions before running Firebase emulators.

## Notes
Closes #259 #199 #14

xgby commented on app/src/androidTest/java/com/neptune/neptune/model/ProfileRepositoryFirebaseTest.kt:
```diff
@@ -264,7 +264,7 @@ class ProfileRepositoryFirebaseTest {
-    repo.observeProfile().test {
+    repo.observeCurrentProfile().test {
       // first emission: should be null (no doc)
       assertNull("fist emission should be null when doc is missing", awaitItem())
```
there's a typo: it should be "first emission should be null when doc is missing"

xgby commented on app/src/main/java/com/neptune/neptune/model/profile/ProfileRepositoryFirebase.kt:
```diff
@@ -160,7 +199,7 @@ class ProfileRepositoryFirebase(
    */
   override suspend fun isUsernameAvailable(username: String): Boolean {
-    val uid = Firebase.auth.currentUser?.uid ?: error("No authenticated user")
+    val uid = auth.currentUser?.uid ?: error("No authenticated user")
```
You have create the helper function requireCurrentUid(). You should call this function.

xgby commented on app/src/main/java/com/neptune/neptune/ui/profile/ProfileScreen.kt:
```diff
@@ -198,22 +199,33 @@ sealed interface ProfileViewConfig {
+              horizontalAlignment = Alignment.CenterHorizontally) {
+                Button(
+                    onClick = onFollow,
+                    enabled = true,
```
There is a delay between the moment I click "Follow" and when the button text changes to "Unfollow." You should disable the button upon clicking it until the state has been updated.

abaur3 commented:
Good catches! Thank you @xgby for our review! I corrected what you suggested, namely: 
- Replace `TODO()` with `UnsupportedOperationException` in test fakes; I think this is a better approach than having a stub implementation. These methods are not supported by the test fake, so by using `throws`, if we'll call it by mistake we'll immediately notice a problem and won't get silent test failures.
- Fix the small typo in the assertion
- Replace code duplication in `isUsernameAvailable` with call to helper method `requireCurrentUid`
- Add a UI state for disabling the follow/unfollow button: when we click on the button and we're waiting for the operation to complete, the button is now disabled and cannot be clicked again during this operation

Moreover, you were right to point out the fact that one could follow himself. It was a bug that I honestly didn't notice and didn't think about. I modified my implementation by adding an `isCurrentUser` helper in MainViewModel to centralise this logic and by updating the navigation: now, when the user clicks on its own avatar icon on a sample that he posted, the app directly navigates to the Self profile screen, preventing a user from following themselves. 

I finally assessed the anonymous user problems: an anonymous user shouldn't be able to Follow anyone, nor to edit their own profile, and other users shouldn't be able to follow the anonymous ones. I implemented this by doing the following:
- Added `isAnonymousUser` to profile UI state and data models; set it when creating a profile for an anonymous user.
- Early-return in `onFollow` when user is anonymous, added `canFollowTarget` flag, and tests for disabled Follow button.
- Defaulted `OtherProfileUiState` to `isAnonymousUser = true` to avoid flashing the Follow button before Firestore loads real data.
- Unified anonymity detection in ViewModels using: `auth.currentUser?.isAnonymous == true || profile.isAnonymous == true`.
- Added tests for anonymous-auth users with no profile and for viewing anonymous profiles.

Let me know if there's still something to be fixed, especially considering that I added many other things after your first review. If something is unclear, don't hesitate to point it out and ask me!
